src/hotspot/share/jfr/utilities/jfrThreadIterator.hpp

```diff
@@ -47,15 +47,17 @@ class JfrThreadIterator : public AP {
 
 class JfrJavaThreadIteratorAdapter {
  private:
-  JavaThreadIteratorWithHandle _iter;
-  JavaThread* _next;
+  ThreadsListHandle _tlist;
```

Comment:
Why do we need to store this?

It looks very suspiocious to have a member that is a stackObj, in a class that is not itself a stackObj. ??

Reply:
The _tlist is used locally in JfrJavaThreadIteratorAdapter constructor only, so it is possible to get rid of it for the price of complicating the constructor a little bit.

Reply:
The `_tlist` is a handle that ensures the captured ThreadsList remains live while the iterator (or a copy) exists.  Dropping it would leave `_it` and `_end` potentially dangling.

Reply:
Okay I had assumed there was a ThreadsList/Handle in the enclosing scope that was being used to initialize this and keep things live, but that is not the case.
I still think it makes no sense to have a StackObj subtype as a member though. Maybe ThreadsListHandle should no longer be a StackObj ?

Reply:
This isn't a change wrto StackObj usage. This adapter class used to have a member that was a JavaThreadIteratorWithHandle (the WithHandle class for the same reason the new code has a ThreadsListHandle, to keep the associated ThreadsList alive), which also derives from StackObj.

Personally, I want to agree with you. I would (and do) use StackObj far more sparingly than it appears in our code base. But I've had this same argument with other folks. The de facto situation, so far as I understand it, is that deriving from StackObj documents normal usage. It has no operational behavior or additional semantics, other than attempting to prevent heap allocation (and it doesn't really succeed there, since a derived class could override that behavior, though I don't currently know of any that do so).  That seems to be what some people want from it and think is useful, and I've stopped trying to convince them otherwise.

Reply:
The adapter is only used as a stack-allocated object, so it can be decorated with StackObj to better communicate this intent.

Reply:
Ah I didn't realise JTIWH was also stackObj.

The thing with StackObj to me is that it indicates a local object that does something interesting on construction and then also on destruction (eg. like MutexLocker). So having one as a member where the destruction is not related to a stack scope seems odd to me. But a stackObj member in a stackObj class seems more reasonable.

Reply:
StackObj to me communicates intent. I don't have to look for "new SomeClass" to see how it is allocated and deallocated, since the new and delete operator are private and won't compile.  Same is true for AllStatic.  I'm one that hasn't been convinced otherwise.  I'm not totally intransigent though as i approved the removal of ValueObj. So there's that.